src/scala/META-INF/native-image/org/pantsbuild/zinc/compiler/README.md

```diff
@@ -0,0 +1,41 @@
+native-image `META-INF/`
+===========================
+
+*This directory contains special configuration files recognized by the `native-image` tool when embedded in a jar. This embedded configuration allows any user with some version of the Graal VM to fetch the `org.pantsbuild:zinc-compiler` jar and run `native-image -jar` to produce a native executable of the pants zinc wrapper without any additional arguments.*
+
+The Graal VM's `native-image` tool[^1] converts JVM bytecode to a native compiled executable[^2] executing via the Substrate VM. The `native-image` tool is not yet immediately compatible with all JVM code[^3], but the tool is stable and featureful enough to successfully build many codebases with a mixture of (mostly) automated and (some) manual configuration.
+
+The `native-image` tool accepts JSON configuration files which overcome some of the current limitations[^3]:
+1. `reflect-config.json`: The largest file, this would typically be generated automatically[^5].
```

Comment:
Can you add the literal command lines you used to generate these files, so that I can re-generate them if I want to by copying and pasting those exact command lines? (And so that I can easily work out what manual edits were made, by running `diff` :))

Reply:
I would need to regenerate these files by running the java agent again on linux, as the reflect configuration currently contains:
1. output from a run of `zinc -help` with the java agent from a previous graal release
2. manual edits
3. output from a run of zinc as used in a pants compile command line with the java agent from the current graal release

(3) was created by running `./pants -ldebug binary zinc:compiler`, stealing one of the command lines, then adding the java agent to the command line. Some of the entries in the `reflect-config.json` which were produced from the runs with the java agent had to be removed, as they caused failures (`NoClassDefFound`) when building native-image. See this comment on the graal PR required to make building zinc work: https://github.com/oracle/graal/pull/1071#issuecomment-473495121 -- the java agent is "still in an experimental state", and the output does change across versions, and while `jq --sort-keys .` will sort keys of individual objects, it doesn't sort different entries of a JSON array in the output, and the output of the `native-image-configure` tool is not guaranteed to be deterministically sorted.

Also, there's no guarantee we're exercising all the right logic paths -- see https://github.com/oracle/graal/blob/master/substratevm/CONFIGURE.md:
> It is advisable to manually review the generated configuration files. Because the agent observes only code that was executed, the resulting configurations can be missing elements that are used in other code paths. It could also make sense to simplify the generated configurations to make any future manual maintenance easier.

If we were to attempt to describe an automated approach, I think we would need to change the approach to generating these to use the environment variable approach mentioned in https://github.com/oracle/graal/blob/master/substratevm/CONFIGURE.md to invoke zinc exactly the way pants is, and then to use the `native-image-configure` tool on the output generated from that.

If the result works, then great. But I'm not sure that given the current instability of the `native-image-configure` tool and surrounding code that it would be worth the time to create that harness right now. I *can* provide the list I gave above on how this was generated, at least.

Reply:
The thing I'm scared of here is that if we need to re-generate them (say, graal fixes a bug, or we update zinc, or we pick up a new compiler), it sounds like it would be hard for _you_ to re-make these files, let alone for _me_ to... And it also sounds like I would need to re-work-out all of the lessons you already learnt to do the manual changes, because I can't know what's different because graal changed vs what's different because you manually modified it.

I'm not saying that I need a script that produces the files exactly as-is, but a script that I can run which does everything except the manual tweaks you made, so that I can run it and run `git diff` against what's checked in to see what you manually did, would be really really useful.

It sounds like this would take the form roughly of an invocation to the java agent tool (at a particular version), with a single zinc command line invocation, at a particular pants sha, and then piping that to `jq --sort-keys .`? Can we try to write down that command line?

Reply:
Yes -- and because I've found that the java agent which generates these files actually works great on OSX (even though it says it's only supported for Linux), and also that it supports the `config-merge-dir` option (which automates the process of generating a config across multiple invocations), producing these command lines will be significantly easier than I expected.